doc/dev/cephfs-reclaim.rst

```diff
@@ -0,0 +1,74 @@
+CephFS Reclaim Interface
+========================
+
+Introduction
+------------
+NFS servers typically do not track ephemeral state on stable storage. If the
+NFS server is restarted, then it will be resurrected with no ephemeral state,
+and the NFS clients are expected to send requests to reclaim what state they
+held during a grace period.
+
+In order to support this use-case, libcephfs has grown several functions that
+allow a client that has been stopped and restarted to destroy or reclaim state
+held by a previous incarnation of itself.
+
+Setting the UUID
+----------------
+In order to properly reset or reclaim against the old session, we need a way
+to identify the old session. This done by setting a unique opaque value on
+the session using **ceph_set_uuid()**. The uuid value can be any string and
+is treated as opaque by the client.
+
+Setting the uuid directly can only be done on a new session, prior to
+mounting. When reclaim is performed the current session will inherit the old
+session's uuid.
+
+Starting Reclaim
+----------------
+After calling ceph_create and ceph_init on the resulting struct
+ceph_mount_info, the client should then issue ceph_start_reclaim, passing in
+the uuid of the previous incarnation of the client with any flags.
+
+CEPH_RECLAIM_RESET
+   This flag indicates that we do not intend to do any sort of reclaim against
+   the old session indicated by the given uuid, and that it should just be
+   discarded. Any state held by the previous client should be released
+   immediately.
+
+Finishing Reclaim
+-----------------
+After the client has completed its reclaim reclaim is complete, the client
+should issue ceph_finish_reclaim to indicate that the reclaim is now
+complete.
+
+Setting Session Timeout (Optional)
+----------------------------------
+When a client dies and is restarted, and we need to preserve its state, we are
+effectively racing against the session expiration clock. In this situation we
+generally want a longer timeout since we expect to eventually kill off the old
+session manually.
+
+Example 1: Reset Old Session
+----------------------------
+(Note: error handling omitted for clarity):
+
+.. code-block:: c
+
+	struct ceph_mount_info *cmount;
+	const char *uuid = "foobarbaz";
+
+	/* Set up a new cephfs session */
+	rc = ceph_create(&cmount);
+	rc = ceph_init(&cmount);
+
+	/* Set the timeout to 5 minutes to lengthen reclaim window */
+	ceph_set_session_timeout(cmount, 300);
+
+	/* Start reclaim vs. session with old uuid */
+	rc = ceph_start_reclaim(cmount, uuid, CEPH_RECLAIM_RESET);
+
```

Comment:
What does it look like to reclaim state at this point? Can you give an example?

Reply:
It's a little hard, if only because there are so many steps involved in doing an open with the ll interfaces. I did add in a ceph_mount() call and a comment mentioning that you can do "normal" open/lock operations at that point for reclaim.